lib/active_model/validations/associated_validator.rb

```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+module ActiveModel
+  module Validations
+    # Borrowed from ActiveRecord: https://github.com/rails/rails/blob/90a1eaa1b30ba1f2d524e197460e549c03cf5698/activerecord/lib/active_record/validations/associated.rb
+    class AssociatedValidator < ActiveModel::EachValidator
+      def validate_each(record, attribute, value)
+        context = record.validation_context.presence
+
+        return if Array(value).reject { |association| valid_object?(association, context) }.none?
+
+        record.errors.add(attribute, :invalid, **options, value:)
+      end
+
+      private
+
+      def valid_object?(record, context)
+        (record.respond_to?(:marked_for_destruction?) && record.marked_for_destruction?) || record.valid?(context)
+      end
+    end
+  end
+end
```

Comment:
Re-use ActiveRecord validations for nested ("associated") models and adapt it to ActiveModel. This lets us remove custom validation logic from the NestedAttributes module.